options.go

```diff
@@ -17,77 +17,420 @@
+					return nil, dbParameters{}, err
+				}
+			}
+
+			if val := pop(query, "dbname"); database == "" {
+				database = val
+			}
+
+			if val := pop(query, "database"); database == "" {
+				database = val
+			}
+
+			if val := pop(query, "user"); user == "" {
+				user = val
+			}
+
+			if val := pop(query, "password"); password == "" {
+				password = val
+			}
+
+			for k, v := range query {
+				serverSettings[k] = v
+			}
+		}
+	} else if dsn != "" {
+		// todo is there a reasonable way to check if name is a python identifier?
+		// nolint https://github.com/edgedb/edgedb-python/blob/325254987314ec7d120cff7ec30be949ec663bc9/edgedb/con_utils.py#L199
+		usingCredentials = true
+
+		u, err := usr.Current()
+		if err != nil {
+			return nil, dbParameters{}, err
+		}
+
+		file := path.Join(u.HomeDir, ".edgedb", "credentials", dsn+".json")
+		data, err := ioutil.ReadFile(file)
+		if err != nil {
+			return nil, dbParameters{}, fmt.Errorf(
+				"cannot read credentials of instance %q: %v%w",
+				dsn, err, ErrClientMistake)
+		}
+
+		var creds credentials
+		if e := json.Unmarshal(data, &creds); e != nil {
+			return nil, dbParameters{}, fmt.Errorf(
+				"cannot read credentials of instance %q: %v%w",
+				dsn, e, ErrClientMistake)
+		}
+
+		if len(ports) == 0 {
+			ports = []int{creds.Port}
+		}
+
+		if user == "" {
+			user = creds.User
+		}
+
+		if len(hosts) == 0 && creds.Host != "" {
+			hosts = []string{creds.Host}
+		}
+
+		if password == "" {
+			password = creds.Password
+		}
+
+		if database == "" {
+			database = creds.Database
+		}
 	}
 
-	var port int
-	if parsed.Port() == "" {
-		port = 5656
-	} else {
-		port, err = strconv.Atoi(parsed.Port())
+	var err error
+
+	if spec := os.Getenv("EDGEDB_HOST"); len(hosts) == 0 && spec != "" {
+		hosts, ports, err = parseHostList(spec, ports)
 		if err != nil {
-			return opts, err
+			return nil, dbParameters{}, err
 		}
 	}
 
-	host := strings.Split(parsed.Host, ":")[0]
-	db := strings.TrimLeft(parsed.Path, "/")
-	password, _ := parsed.User.Password()
+	if len(hosts) == 0 {
+		if !usingCredentials {
+			hosts = append(hosts, defaultHosts...)
+		}
+		if !opts.admin {
+			hosts = append(hosts, "localhost")
+		}
+	}
 
-	return Options{
-		Host:     host,
-		Port:     port,
-		User:     parsed.User.Username(),
-		Database: db,
-		Password: password,
-	}, nil
+	if len(ports) == 0 {
+		if portSpec := os.Getenv("EDGEDB_PORT"); portSpec != "" {
+			ports, err = parsePortSpec(portSpec)
+			if err != nil {
+				return nil, dbParameters{}, err
+			}
+		} else {
+			ports = []int{5656}
+		}
+	}
+
+	ports, err = validatePortSpec(hosts, ports)
+	if err != nil {
+		return nil, dbParameters{}, err
+	}
+
+	if val := os.Getenv("EDGEDB_USER"); user == "" {
+		user = val
+	}
+
+	if val := os.Getenv("EDGEDB_PASSWORD"); password == "" {
+		password = val
+	}
+
+	if val := os.Getenv("EDGEDB_DATABASE"); database == "" {
+		database = val
+	}
+
+	if database == "" {
+		database = user
+	}
+
+	if user == "" {
```

Comment:
The default user/database should be "edgedb" -- see https://github.com/edgedb/edgedb-python/pull/131